crates/metaslang/cst/src/cst.rs

```diff
@@ -1,67 +1,65 @@
-// This file is generated automatically by infrastructure scripts. Please don't edit by hand.
-
 use std::rc::Rc;
 
 use serde::Serialize;
 
 use crate::cursor::Cursor;
-use crate::kinds::{NodeLabel, RuleKind, TokenKind};
 use crate::text_index::TextIndex;
+use crate::{KindTypes, TerminalKind as _};
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
-pub struct LabeledNode {
-    pub label: Option<NodeLabel>,
-    pub node: Node,
-}
-
-impl LabeledNode {
-    /// Creates an anonymous node (without a label).
-    pub fn anonymous(node: Node) -> Self {
-        Self { label: None, node }
-    }
-}
-
-impl std::ops::Deref for LabeledNode {
-    type Target = Node;
-
-    fn deref(&self) -> &Self::Target {
-        &self.node
-    }
+pub struct TerminalNode<T: KindTypes> {
+    pub kind: T::TerminalKind,
+    pub text: String,
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
-pub struct RuleNode {
-    pub kind: RuleKind,
+pub struct NonTerminalNode<T: KindTypes> {
+    pub kind: T::NonTerminalKind,
     pub text_len: TextIndex,
     #[serde(skip_serializing_if = "Vec::is_empty")]
-    pub children: Vec<LabeledNode>,
+    pub children: Vec<LabeledNode<T>>,
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
-pub struct TokenNode {
-    pub kind: TokenKind,
-    pub text: String,
+pub enum Node<T: KindTypes> {
+    Rule(Rc<NonTerminalNode<T>>),
+    Token(Rc<TerminalNode<T>>),
 }
 
 #[derive(Clone, Debug, PartialEq, Eq, Serialize)]
-pub enum Node {
-    Rule(Rc<RuleNode>),
-    Token(Rc<TokenNode>),
+pub struct LabeledNode<T: KindTypes> {
+    pub label: Option<T::EdgeKind>,
+    pub node: Node<T>,
+}
+
+impl<T: KindTypes> LabeledNode<T> {
+    /// Creates an anonymous node (without a label).
+    pub fn anonymous(node: Node<T>) -> Self {
+        Self { label: None, node }
+    }
+}
+
+impl<T: KindTypes> std::ops::Deref for LabeledNode<T> {
+    type Target = Node<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.node
+    }
 }
 
-impl Node {
-    pub fn rule(kind: RuleKind, children: Vec<LabeledNode>) -> Self {
+impl<T: KindTypes> Node<T> {
+    pub fn rule(kind: T::NonTerminalKind, children: Vec<LabeledNode<T>>) -> Self {
         let text_len = children.iter().map(|node| node.text_len()).sum();
 
-        Self::Rule(Rc::new(RuleNode {
+        Self::Rule(Rc::new(NonTerminalNode::<T> {
             kind,
             text_len,
             children,
         }))
     }
 
-    pub fn token(kind: TokenKind, text: String) -> Self {
-        Self::Token(Rc::new(TokenNode { kind, text }))
+    pub fn token(kind: T::TerminalKind, text: String) -> Self {
+        Self::Token(Rc::new(TerminalNode::<T> { kind, text }))
```

Comment:
ditto here and in all of the similar places

Reply:
as above, so below